server/proxy/tcphttptunnel.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2020 guylewin, guy@lewin.co.il
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"github.com/fatedier/frp/models/config"
+	"github.com/fatedier/frp/utils/util"
+	"github.com/fatedier/frp/utils/vhost"
+	"strings"
+)
+
+type TcpHttpTunnelProxy struct {
+	*BaseProxy
+	cfg *config.TcpHttpTunnelProxyConf
+
+	realPort int
+}
+
+func (pxy *TcpHttpTunnelProxy) Run() (remoteAddr string, err error) {
+	xl := pxy.xl
+	routeConfig := &vhost.VhostRouteConfig{}
+
+	defer func() {
+		if err != nil {
+			pxy.Close()
+		}
+	}()
+	addrs := make([]string, 0)
+	for _, domain := range pxy.cfg.CustomDomains {
```

Comment:
Inside this `for` you start listening in several ports. If the function succeeds all is good, but what happens if you `return` on an error? Shouldn't you clean up after you on all listeners that you started?

Reply:
If there's an error (`err != nil`) `pxy.Close()` will be called, which calls l.Close() on every listener in the array.